sympy/physics/mechanics/joint.py

```diff
@@ -0,0 +1,681 @@
+from sympy import acos, Symbol
+from sympy.physics.vector import cross, Vector, dot, dynamicsymbols
+from sympy.physics.mechanics.functions import convert_tuple_to_vector
+
+__all__ = ['Joint', 'PinJoint', 'SlidingJoint', 'CylindricalJoint',
+           'SphericalJoint', 'PlanarJoint']
+
+
+class Joint(object):
+    """Abstract Base class for all specific joints.
+
+    A Joint connects two bodies (a parent and child) by adding different degrees
+    of freedom to child with respect to parent.
+    This is the base class for all specific joints and holds all common methods
+    acting as an interface for all joints. Custom joint can be created by
+    creating a subclass of this class and overriding 'apply_joint()' method to
+    add the dynamics of the joint.
+
+    Parameters
+    ----------
+    name: String
+        Name of the joint which makes it unique. Should be different from other
+        joints.
+    parent: Body
+        Instance of Body class serving as the parent in the joint.
+    child: Body
+        Instance of Body class serving as the child in the joint.
+    parent_point_pos: 3 Tuple, Optional
+        Defines the joint's point where the parent will be connected to child.
+        3 Tuple defines the values of x, y and z directions w.r.t parent's
+        frame. If it is not supplied, center of mass is added as default.
+    child_point_pos: 3 Tuple, Optional
+        Defines the joint's point where the child will be connected to parent.
+        3 Tuple defines the values of x, y and z directions w.r.t child's frame.
+        If it is not supplied, center of mass is added as default.
+    parent_axis: Vector, Optional
+        Defines the orientation as a vector which must be aligned with child's
+        axis before adding joint. If it is not passed, default is x axis in
+        parent's frame.
+    child_axis: Vector, Optional
+        Defines the orientation as a vector which must be aligned with parent's
+        axis before adding joint. If it is not passed, default is x axis in
+        child's frame.
+    """
+
+    def __init__(self, name, parent, child, parent_point_pos=None,
+                 child_point_pos=None):
+        self.name = name
+        self.parent = parent
+        self.child = child
+        self.coordinates = []
+        self.speeds = []
+        self.kds = []
+
+        if parent_point_pos is None:
+            parent_point_pos = (0, 0, 0)  # Parent's Center of mass
+
+        if child_point_pos is None:
+            child_point_pos = (0, 0, 0)  # Child's Center of mass
+
+        self._parent_joint_location = convert_tuple_to_vector(
+            parent.frame, parent_point_pos)
+        self._child_joint_location = convert_tuple_to_vector(
+            child.frame, child_point_pos)
+
+        self._locate_joint_point()
+        self.apply_joint()
```

Comment:
@moorepants Instead of calling `apply_joint` internally , I think it would be better to make user call `apply_joint` with degrees of freedom as parameters.

Like for Pin Joint it can be-

```
>>> theta, omega = dynamicsymbols(theta, omega)
>>> parent = Body('parent')
>>> child = Body('child')
>>> child.apply_force()
>>> P = PinJoint('P', parent, child)
>>> P.apply_joint(theta, omega)

```

Reply:
What about:

```
PinJoint.__init__(self, name, parent, child, coordinate=None, speed=None)
```

If they are set to None, we auto generate them. If we autgenerate them, they have to be unique for a collection of Joints. Which means they can't be autogenerated until provided to JointMethod unless we have a global way to track them. Or maybe the pinjoint name is what makes them unique. So if every pin join has PinJoint.coord == Function('theta{}'.format(PinJoint.name)(t) then the uniqueness of the varialbes is controlled by the pin joint's name.

Reply:
> What about:
> 
> ```
> PinJoint.__init__(self, name, parent, child, coordinate=None, speed=None)
> ```

I think coordinate and speed shouldn't go in `Joint.__init__` . It should be something like

```
P = PinJoint('P', parent, child)
P.apply_joint(coordinate=None, speed=None)

soln = JointsMethod(ground, P)
```
User should use `apply_joint()` with coordinates and speed parameters, before passing Joint to JointsMethod.

> Or maybe the pinjoint name is what makes them unique. So if every pin join has `PinJoint.coord == Function('theta{}'.format(PinJoint.name)(t)` then the uniqueness of the varialbes is controlled by the pin joint's name.

I think uniqueness controlled by joint's name would be good.

Reply:
Seems reasonable. Show it in examples. It is hard to see how everything will fit together without drafting up examples first. I highly recommend taking several examples from easy to hard ones in the sympy and pydy repos to demonstrate the new API. Putting those in your GSoC proposal is the best way to communicate what you want to do. These examples then will become tests, either in the unit tests or as auto tested documentation.

Reply:
Yes sure! Thanks!!

Reply:
@moorepants I have added tests and proposed API to my GSoC proposal. Please review my GSoC proposal [Sudeep Sidhu GSoC Proposal](https://docs.google.com/document/d/1dRPdFcJqk0BDDjNcaCvOSDaCo6BCkZV1c8HJYMIogKw/edit?usp=sharing)

Reply:
@sidhu1012 I don't plan to review any proposals before submission. If you have specific questions you can ask them here or the mailing list and I will answer if I have time.

Reply:
Okay

Reply:
@moorepants I have submitted by draft proposal on official GSoC page.